Move game speed definitions to JSON files

OptimizedForDensity opened the pull request:
This PR moves the definition of game speeds from Kotlin source files to JSON files. New game speed variables are also ported from Civ 5 and integrated in ways as close to the original implementation as I could find. These changes should allow for greater flexibility and modability of game speeds and game speed modifiers with respect to certain game mechanics (e.g., slow down production speed without changing research speed). New game speeds can be easily added as seen here

![Example of a modded game speed](https://files.catbox.moe/ec7tcj.jpg)

Modders can also modify how many years each turn takes so you could, for example, make a World War II mod where each calendar year has 52 turns (1 week per turn).

Gameplay changes introduced by this PR:
* Golden ages are longer on quick and shorter on epic and marathon
* Peace treaties last longer on epic and marathon

Todo:

- [ ] Clean up last few references to the old game speed file
- [ ] Modding documentation
- [ ] Civilopedia entries for game speeds

Since this is such a fundamental change I decided to open it up as a draft for discussion and feedback. The biggest thing I've yet to decide on is if any of the individual speed modifiers should be removed. Some of these variables I couldn't find used in source code or are used in ways I didn't expect (Civ 5 city state quest cooldowns are scaled according to ```greatPeoplePercent```) and in most cases these speed modifiers could probably be represented by a general speed ```modifier``` as originally coded in Unciv (which I've kept in game speed definitions and in some places in code). If it's desirable to keep all or the majority of individual speed modifiers for the sake of modability I can keep them as I've coded them, if not I can start replacing them with ```modifier```.

yairm210 commented:
A. There's a lot of detail here that looks like pre-optimizing. Who said any modders want to divide between percentages for research, gold, improvements, etc? All that is currently contained in a single parameter and that's Good Enough

B. Default game speeds may not exist, so we have to account for that when defaulting (e.g. the final change which has GameSpeed as a string)

C. What happens if a mod *doesn't* want custom speeds? I don't see you using the default game speeds, which is what I as a modder would expect (we do this for a bunch of other things as well - undefined means 'use game defaults')

SomeTroglodyte commented on core/src/com/unciv/logic/GameInfo.kt:
```diff
@@ -129,6 +131,7 @@ class GameInfo {
      *  @throws NoSuchElementException in no-barbarians games! */
     fun getBarbarianCivilization() = getCivilization(Constants.barbarians)
     fun getDifficulty() = difficultyObject
+    fun getGameSpeed() = gameSpeedObject
```
If there's an accessor, make the backing field private?

OptimizedForDensity replied:
Done

SomeTroglodyte commented on core/src/com/unciv/models/ruleset/Ruleset.kt:
```diff
@@ -81,6 +81,7 @@ class Ruleset {
     val buildings = LinkedHashMap<String, Building>()
     val difficulties = LinkedHashMap<String, Difficulty>()
     val eras = LinkedHashMap<String, Era>()
+    val speeds = LinkedHashMap<String, GameSpeed>()
```
I'd like gameSpeeds marginally better

OptimizedForDensity replied:
Done

SomeTroglodyte commented:
Finally - getting rid of all those `when (gameInfo.gameParameters.gameSpeed) {` is a worthy goal indeed. Looking forward to a final draft.

OptimizedForDensity commented on desktop/src/com/unciv/app/desktop/ConsoleLauncher.kt:
```diff
@@ -67,7 +67,7 @@ internal object ConsoleLauncher {
         return GameParameters().apply {
             difficulty = "Chieftain"
-            gameSpeed = GameSpeed.Quick
+            gameSpeed = "Quick"
```
Good idea, I went ahead with those added those. Since I saw quick and standard both used in code I added const strings ```GameSpeed.DEFAULT``` and ```GameSpeed.DEFAULTFORSIMULATION```

OptimizedForDensity commented:
> A. There's a lot of detail here that looks like pre-optimizing. Who said any modders want to divide between percentages for research, gold, improvements, etc? All that is currently contained in a single parameter and that's Good Enough

I went ahead and reduced everything down to a general ```modifier``` then also added in durations or other modifiers that were contained within ```when``` blocks (```barbPercent```, ```peaceDuration```, ```goldGiftMod```, ```religiousPressureAdjacentCity```, and ```dealDuration```). ```goldenAgePercent``` (which probably should be renamed at this point along with ```barbPercent``` tbh) remains since that doesn't scale with ```modifier```.

> B. Default game speeds may not exist, so we have to account for that when defaulting (e.g. the final change which has GameSpeed as a string)

I still need to fully work this one out but that's a good point.

> C. What happens if a mod doesn't want custom speeds? I don't see you using the default game speeds, which is what I as a modder would expect (we do this for a bunch of other things as well - undefined means 'use game defaults')

Good catch, I added a step that catches if no game speeds are given and substitutes in the default game speeds.

SomeTroglodyte commented:
> Good catch, I added a step that catches if no game speeds are given and substitutes in the default game speeds

Not having looked at the code: We have a vanilla json, and if an extension mod doesn't supply its own, no worries. If a base mod doesn't we would normally not fall back to vanilla, right? But doing so for this file would make sense. With a modcheck warning. Lastly, if a mod supplied `[]` in its file, we could crash, to prevent that - modcheck hard-error (that is making sure NewGameScreen refuses) or insert _one_ default speed called "None" ? Just a little brainstorm, feel free to ignore.

SomeTroglodyte commented on core/src/com/unciv/ui/civilopedia/CivilopediaCategories.kt:
```diff
@@ -188,6 +188,11 @@ enum class CivilopediaCategories (
+    GameSpeed ("Game Speeds", false,
+        getImage = null,
+        KeyCharAndCode('S'),
+        "EmojiIcons/Turn"
```
One of the textures going in at 50x50, the rest is bigger. Probably just about visible. Your choice if that displeases you. Can always be changed later (actually all the category icons should be independently moddable and not reuse names used elsewhere, the packer will merge identical copies as long as they're in the same atlas pack - me too I was too lazy often for new uses)

OptimizedForDensity replied:
I see now that now. I've gotten a new (attributed, royalty-free) icon uploaded for use as the game speed civilopedia icon and anywhere else that someone sees fit.

SomeTroglodyte replied:
Sorry for being too terse before. If you looked up the hourglass in credits.md and re-downloaded that as base, likely fine, maybe mention the new use. Any other Art must be attributed in said credits.md file, _with_ license type. Your new entry is likely within some "unless otherwise stated everything is Creative Commons xyz" scope, so please check whether your source has a clear license policy and whether that matches. All licenses must be compatible with our own license, which I forget what it is. All very rigid thinking, annoying in a way, but a project like this needs to abide by the powers.

SomeTroglodyte replied:
Oh, and not your fault, but mine - the auto-reformatting the editor did is probably wrong. Try to copy the markdown into any github markdown preview and compare... If it renders the same, fine, if not, try to revert. And tell me. Could go into `.editorconfig` if I find the proper keywords.

SomeTroglodyte commented on core/src/com/unciv/models/ruleset/GameSpeed.kt:
```diff
@@ -22,6 +25,16 @@ class GameSpeed : RulesetObject() {
     override fun makeLink(): String = "GameSpeed/$name"
+    override fun getCivilopediaTextHeader() = FormattedLine(name, header = 2)
+    override fun getCivilopediaTextLines(ruleset: Ruleset) = sequence {
+        yield(FormattedLine("General speed modifier: [${modifier * 100}]% ${Fonts.turn}"))
```
Doesn't quite match your template - this way the Fonts symbol (and the blank) becomes part of the key. Flexible for translators, more  work for us. The opposite would be `"{General speed modifier}: ${modifier * 100}% ${Fonts.turn}"` and just "General speed modifier" as translation key, but xlators would have no control over the rest of the line... Your choice, but try a quick test with a lang of your choice.

OptimizedForDensity replied:
I haven't had the fortune of working with translations yet. Are my changes just now what you were meaning? I tried the new translation.properties with a French translation of the new strings and it seems to work.

OptimizedForDensity commented:
The PR is now at a point where I consider it a minimally-acceptable release state (barring any bugs that I haven't caught but I tried to be thorough with my testing; base ruleset mods without any speeds will have the default speeds added and any default speed declarations are adjusted to prevent crashes if the default speeds reference now-deleted speed objects). As far as adding other speeds are concerned, I think good starting points for more speed modifiers would be
* research
* culture cost
* (general) construction cost (tile improvements, buildings, units, great person hurry contribution)
* gold cost modifiers

Other more specific ones could include
* city state quest timers
* great person modifiers

I think starting with the first 4 could be straight forward enough to understand their impact (unlike some Civ 5 speed modifiers) while giving modders flexibility in slowing or speeding up any aspects of gameplay.

SomeTroglodyte reviewed (approved):
I play-tested, runs just fine, Marathon slows things down as supposed to...

yairm210 commented on core/src/com/unciv/logic/civilization/CityStateFunctions.kt:
```diff
@@ -393,17 +388,13 @@ class CityStateFunctions(val civInfo: CivilizationInfo) {
-            GameSpeed.Marathon -> 32f
+        var gold = (10 * civInfo.gameInfo.getGameSpeed().goldGiftModifier).toInt() * 5 // rounding down to nearest 5
+        val turnsToIncrement = when (civInfo.gameInfo.getGameSpeed().name) {
+            "Quick" -> 5f
+            "Standard" -> 6.5f
+            "Epic" -> 14f
+            "Marathon" -> 32f
```
Good catch :D

yairm210 commented:
What happens if you add a mod, select a moddable game speed, then remove that mod?

OptimizedForDensity commented on core/src/com/unciv/logic/city/CityReligion.kt:
```diff
@@ -4,40 +4,32 @@ import com.unciv.Constants
-    
-    var religionThisIsTheHolyCityOf: String? = null 
-    
+    private val pressureFromAdjacentCities: Int by lazy { cityInfo.civInfo.gameInfo.getGameSpeed().religiousPressureAdjacentCity }
```
`cityInfo` is a `lateinit` variable in this class to avoid a cyclic reference (`CityInfo()` calls this class' constructor) so `lazy` is needed.

SomeTroglodyte replied:
Not sure what your "Wtf" is referring to... But with "shit" and "pressure" involved I guess something with diarrhoea? Well, that's religion for you.

OptimizedForDensity commented on android/assets/jsons/Civ V - Gods & Kings/GameSpeeds.json:
```diff
@@ -0,0 +1,141 @@
+[
+    // Source: https://github.com/Eyefyre/Civilization-5-Data/blob/main/XML/Expansion_1/GameInfo/CIV5GameSpeeds.xml
```
That's sad, they were a good source of Civ 5 XML data.

SomeTroglodyte replied:
No [Wayback](https://web.archive.org/web/*/https://github.com/Eyefyre/Civilization-5-Data/blob/main/XML/Expansion_1/GameInfo/CIV5GameSpeeds.xml) either. Perhaps we should start _triggering_ a wayback archive for such links while they're still alive, and use those right away?

OptimizedForDensity commented on core/src/com/unciv/logic/GameInfo.kt:
```diff
@@ -132,6 +136,12 @@ class GameInfo {
+        gameSpeedObject = newGameSpeed
+        gameParameters.gameSpeed = newGameSpeed.name
+    }
+
```
I realized that the custom setter isn't even needed since starting a new game (or loading a saved one) goes something like
1. set `gameParameters.speed` using dropdown menu (or read from file)
2. set `speed` object using `ruleSet.speeds[gameParameters.speed]`

OptimizedForDensity commented on core/src/com/unciv/models/ruleset/GameSpeed.kt:
```diff
@@ -0,0 +1,98 @@
+    var dealDuration: Int = 30
+    var turnIncrements: ArrayList<HashMap<String, Float>> = ArrayList()
+
+    val yearsToTurnObject: ArrayList<YearsPerTurn> by lazy { initYearsToTurn(turnIncrements) }
```
De-objectication complete and variables renamed.

OptimizedForDensity commented on core/src/com/unciv/models/ruleset/GameSpeed.kt:
```diff
@@ -0,0 +1,98 @@
+        if (goldGiftModifier < 0f) goldGiftModifier = modifier
+        if (barbarianModifier < 0f) barbarianModifier = modifier
+        if (goldenAgeLengthModifier < 0f) goldenAgeLengthModifier = modifier
+    }
```
I think that error popup comment was a classic case of me starting to type something, stopping for a moment to think, then getting distracted and never finishing it since I don't even know what I was talking about (maybe the check that happens when a ruleset is loaded).

OptimizedForDensity commented on core/src/com/unciv/models/ModConstants.kt:
```diff
@@ -62,6 +62,8 @@ class ModConstants {
     var minRiverLength = 5
     var maxRiverLength = 666  // Do not set < max map radius
 
+    var startYear = -4000f
```
`startYear` moved to Speed.json definitions.

OptimizedForDensity commented:
I've finally gotten all feedback addressed. The biggest change was renaming GameSpeed back to Speed per Azzurite's suggestion.

> What happens if you add a mod, select a moddable game speed, then remove that mod?

If you mean in the new game setup screen, the dropdown box will automatically update the game speeds to one contained in the new combined ruleset. There is a failsafe in the ruleset loading to make sure that there's at least one valid game speed (if not, it will load the vanilla game speeds so a base ruleset mod will work even if the author does not add any game speeds).

OptimizedForDensity commented on android/assets/jsons/Civ V - Gods & Kings/Speeds.json:
```diff
@@ -0,0 +1,113 @@
+        "religiousPressureAdjacentCity": 9,
+        "peaceDealDuration": 10,
+        "dealDuration": 25,
+        "startYear": -4000.0,
```
Just tested and no they don't.

SomeTroglodyte replied:
Afaik Gdx Json will happily poke floats into ints and vice versa. I'd actually _like_  a .0 if it's meant as reminder that the backing val is in fact float...

ravignir commented:
` companion object {
        const val DEFAULT: String = "Quick"
        const val DEFAULTFORSIMULATION: String = "Standard"
    }`
... naisu, you messed that up @OptimizedForDensity

OptimizedForDensity commented:
The default game speed before this PR was "Quick", this companion object just keeps the default string in a central location. Or do you mean something else?

OptimizedForDensity commented on core/src/com/unciv/models/metadata/GameParameters.kt:
```diff
@@ -9,7 +10,7 @@ enum class BaseRuleset(val fullName:String){
 
 class GameParameters { // Default values are the default new game
     var difficulty = "Prince"
-    var gameSpeed = GameSpeed.Standard
```
Ah I was misremembering (there's somewhere else in code it was set `gameSpeed = GameSpeed.Quick` but looking back it was in `ConsoleLauncher.kt`).

calrissian-tendra commented:
Now that  speed settings are moddable, which is a nice feature, I'd like to see them made into uniques, and perhaps even allowed to be set per era as well for even more modability.
I haven't the skill to do that myself, but would love to use that feature in a mod.
